op-node/rollup/derive/batch_queue.go

```
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
+			panic(fmt.Errorf("unknown batch validity type: %d", validity))
 		}
+	}
+	// clean up if we remove the final batch for this timestamp
+	if len(remaining) == 0 {
+		delete(bq.batches, nextTimestamp)
+	} else {
+		bq.batches[nextTimestamp] = remaining
+	}
 
-		// 2b. Determine the valid time window
-		l1OriginTime := bq.l1Blocks[0].Time
-		nextL1BlockTime := bq.l1Blocks[1].Time // Safe b/c the epoch is the L1 Block number of the first block in L1Blocks
-		minL2Time := l2SafeHead.Time + bq.config.BlockTime
-		maxL2Time := l1OriginTime + bq.config.MaxSequencerDrift
-		newEpoch := l2SafeHead.L1Origin != epoch.ID() // Only guarantee a L2 block if we have not already produced one for this epoch.
-		if newEpoch && minL2Time+bq.config.BlockTime > maxL2Time {
-			maxL2Time = minL2Time + bq.config.BlockTime
+	if nextBatch != nil {
+		// advance epoch if necessary
+		if nextBatch.Batch.EpochNum == rollup.Epoch(epoch.Number)+1 {
+			bq.l1Blocks = bq.l1Blocks[1:]
 		}
-
-		bq.log.Trace("found batches", "len", len(batches))
-		// Filter + Fill batches
-		batches = FilterBatches(bq.log, bq.config, epoch.ID(), minL2Time, maxL2Time, batches)
-		bq.log.Trace("filtered batches", "len", len(batches), "l1Origin", bq.l1Blocks[0], "nextL1Block", bq.l1Blocks[1], "minL2Time", minL2Time, "maxL2Time", maxL2Time)
-		batches = FillMissingBatches(batches, epoch.ID(), bq.config.BlockTime, minL2Time, nextL1BlockTime)
-		bq.log.Trace("added missing batches", "len", len(batches), "l1OriginTime", l1OriginTime, "nextL1BlockTime", nextL1BlockTime)
-		// Advance an epoch after filling all batches.
-		bq.l1Blocks = bq.l1Blocks[1:]
-
-		return batches, nil
-
-	} else {
-		bq.log.Trace("Trying to eagerly find batch")
-		next, err := bq.tryPopNextBatch(ctx, l2SafeHead)
-		if err != nil {
-			return nil, err
-		} else {
-			bq.log.Info("found eager batch", "batch", next.Batch)
-			return []*BatchData{next.Batch}, nil
+		// sanity check
+		if nextBatch.Batch.EpochNum > rollup.Epoch(epoch.Number)+1 {
```

This check could be more strict

```
if nextBatch.Batch.EpochNum != rollup.Epoch(epoch.Number)
  && nextBatch.Batch.EpochNum != rollup.Epoch(epoch.Number)+1
```

This would silently fail for going backwards (which I assume isn't allowed). Could also just add another `if` clause that explicitly checks the going backwards case

---

`CheckBatch` already checks these edge-cases. What do you think about removing the sanity check here? It was mostly a product of testing/debugging

---

I'm cool with removing this sanity check if its already checked elsewhere